### PR TITLE
Attribute inference attacks

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

Implemented three attribute inference attacks: one black-box and two white-box attacks for decision trees.

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Added 4 unit tests for the new attacks

**Test Configuration**:
- OS: MacOS mojave 10.14.6
- Python version: 3.7
- TensorFlow 

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/inference/attribute_inference.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2020
```
Comment (mathsinn) [inline art/attacks/inference/attribute_inference.py:3]: Could you please change this line to
```
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
This is the author/copyright convention we recently adopted.

File: tests/attacks/inference/test_attribute_inference.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
+# MIT License
+#
+# Copyright (C) IBM Corporation 2020
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:3]: Could you please change this line to
```
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
This is the author/copyright convention we recently adopted.

File: tests/attacks/inference/test_attribute_inference.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        attack.fit(x_train)
+        # infer attacked feature
+        inferred_train = attack.infer(x_train_for_attack, x_train_predictions, values=values)
+        inferred_test = attack.infer(x_test_for_attack, x_test_predictions, values=values)
+        # check accuracy
+        train_acc = np.sum(inferred_train == x_train_feature.reshape(1,-1)) / len(inferred_train)
+        test_acc = np.sum(inferred_test == x_test_feature.reshape(1,-1)) / len(inferred_test)
+        print(train_acc)
+        print(test_acc)
+        # assert train_acc > test_acc
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:83]: Is this line meant to be commented out?

File: tests/attacks/inference/test_attribute_inference.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:142]: Is this line meant to be commented out?

File: tests/attacks/inference/test_attribute_inference.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+
+def test_white_box(get_tabular_classifier_list, get_iris_dataset):
+    classifier_list = get_tabular_classifier_list(AttributeInferenceWhiteBoxDecisionTree)
+    if not classifier_list:
+        logging.warning("Couldn't perform  this test because no classifier is defined")
+        return
+
+    attack_feature = 2  # petal length
+    # values = [0.14, 0.42, 0.57, 0.71, 0.85] # rounded down
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:152]: Could this line be removed?

File: tests/attacks/inference/test_attribute_inference.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    values = [0.14, 0.42, 0.71]  # rounded down
+    # priors = [50/150, 11/150, 43/150, 35/150, 11/150]
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:154]: Could this line be removed?

File: tests/attacks/inference/test_attribute_inference.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        x_train_predictions = np.array([np.argmax(arr) for arr in classifier.predict(x_train_iris)]).reshape(-1,1)
+        x_test_predictions = np.array([np.argmax(arr) for arr in classifier.predict(x_test_iris)]).reshape(-1,1)
+        inferred_train = attack.infer(x_train_for_attack, x_train_predictions, values=values, priors=priors)
+        inferred_test = attack.infer(x_test_for_attack, x_test_predictions, values=values, priors=priors)
+        train_diff = np.abs(inferred_train - x_train_feature.reshape(1,-1))
+        test_diff = np.abs(inferred_test - x_test_feature.reshape(1,-1))
+        print(np.sum(train_diff) / len(inferred_train))
+        print(np.sum(test_diff) / len(inferred_test))
+        # if type(classifier).__name__ is not 'ScikitlearnDecisionTreeClassifier':
+        # assert np.sum(train_diff) / len(inferred_train) < np.sum(test_diff) / len(inferred_test)
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:175]: Are these two lines meant to be commented out?

File: tests/attacks/inference/test_attribute_inference.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+        attack = AttributeInferenceWhiteBoxLifestyleDecisionTree(classifier, attack_feature=attack_feature)
+        x_train_predictions = np.array([np.argmax(arr) for arr in classifier.predict(x_train_iris)]).reshape(-1,1)
+        x_test_predictions = np.array([np.argmax(arr) for arr in classifier.predict(x_test_iris)]).reshape(-1,1)
+        inferred_train = attack.infer(x_train_for_attack, x_train_predictions, values=values, priors=priors)
+        inferred_test = attack.infer(x_test_for_attack, x_test_predictions, values=values, priors=priors)
+        train_diff = np.abs(inferred_train - x_train_feature.reshape(1,-1))
+        test_diff = np.abs(inferred_test - x_test_feature.reshape(1,-1))
+        print(np.sum(train_diff) / len(inferred_train))
+        print(np.sum(test_diff) / len(inferred_test))
+        # assert np.sum(train_diff) / len(inferred_train) < np.sum(test_diff) / len(inferred_test)
```
Comment (mathsinn) [inline tests/attacks/inference/test_attribute_inference.py:207]: Is this line meant to be commented out?

**Comment by beat-buesser** [comment]
@abigailgold and @mathsinn Thank you very much!

**Pull request merged**